### ISSUE TITLE
failure when try to define and subsequently call a macro in a try block

### ISSUE THREAD

**Issue by ikappaki** [basilisp-lang/basilisp#1086]
Hi,

it doesn't seem to be possible to define and subsequently call a macro inside a try block
```
error occurred during macroexpansion: 'Unbound' object is not callable
```

To reproduce,
open the nREPL and define and call a trivial macro inside a try block, it fails with the above error
```clojure
> basilisp repl
basilisp.user=> (try
                  (defmacro abc []
                    '(println 5))
                  (abc)

                  (catch Exception e
                    e))


  exception: <class 'TypeError'> from <class 'basilisp.lang.compiler.exception.CompilerException'>
      phase: :macroexpansion
    message: error occurred during macroexpansion: 'Unbound' object is not callable
       form: (abc)
   location: <REPL Input>:4
```

The same works fine at the top level
```clojure
> basilisp repl
basilisp.user=> (defmacro abc []
                    '(println 5))
                  (abc)
5
nil
```

Is there a technical constraint perhaps that prevents macros from being defined anywhere else other than the top level?

Thanks

**Comment by chrisrink10** [comment]
Yes, `try` blocks are evaluated as a single compilation unit. I see that this works in Clojure though I'm not sure how. I do not expect this is something that will be easily fixed.

**Comment by chrisrink10** [comment]
I'm confused why this issue and issues like it keep coming up. This is not like any code I find myself typically writing so can you give me some context as to why such dynamic code is necessary?

**Comment by ikappaki** [comment]
> I'm confused why this issue and issues like it keep coming up. This is not like any code I find myself typically writing so can you give me some context as to why such dynamic code is necessary?

Thanks for taking a look. I'm writing a macro for remote Basilisp code execution in another process and encountering errors with syntax that seems valid but throws exceptions. The macro, in its simplest form for the purpose of illustration, wraps the remote code in a `try` block, which can include various forms.

```clojure
(defmacro with-remote-eval [& code]
  `(try
     ~@code
     (catch python/Exception e
       e)))

(with-remote-eval
  code)
```

I consider this a valid macro use case.

I've found a workaround for unsupported forms within `try`, like `require`, `import`, and now `defmacro`. These are moved to a top-level `do` form in the remote execution code, allowing the unsupported forms to work as long as they are allowed in `do` (which they currently are).

```clojure
(defmacro with-remote-eval [& code]
  (let [[code-unsupported code-rest] (split-code code)]
    `(do
       ~@code-unsupported
       (try
         ~@code-rest
         (catch python/Exception e
           e)))))
```

Let me know if these errors aren't worth reporting, and I'll adjust accordingly.

Thanks

**Comment by amano-kenji** [comment]
Defining a macro in a try expression seems like an anti-pattern to me.
